### PR TITLE
Ensure transaction inputs/outputs are sorted lexicographically

### PR DESCRIPTION
**Pull request by Adam-Gleave**
# Description of change

Checks that inputs/outputs of a transaction payload are sorted lexicographically, to be consistent with syntactic validation. If this is in the wrong place, please let me know – still working my way around!

## Type of change

Choose a type of change, and delete any options that are not relevant.

- Bug fix (a non-breaking change which fixes an issue)

## How the change has been tested

Added unit tests to ensure that `is_sorted` is consistent for comparing `Vec`s.

## Change checklist

Tick the boxes that are relevant to your changes, and delete any items that are not.

- [x] I have followed the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that new and existing unit tests pass locally with my changes

### CODE REVIEW

File: bee-message/src/payload/transaction/essence/regular.rs [hunk 1]
```diff
@@ -234,12 +235,25 @@ impl RegularEssenceBuilder {
             }
         }
 
-        // TODO
-        // inputs.sort();
-        // outputs.sort();
+        // Inputs must be lexicographically sorted in their serialised form
+        let packed_inputs: Result<Vec<_>, _> = self.inputs.iter().map(|input| {
+            let mut packed_input = vec![];
+            input.pack(&mut packed_input).and(Ok(packed_input))
```
Comment (thibault-martinez) [inline bee-message/src/payload/transaction/essence/regular.rs]: There is `pack_new` that returns a new filled vector :)

File: bee-message/src/payload/transaction/essence/regular.rs [hunk 2]
```diff
@@ -234,12 +235,25 @@ impl RegularEssenceBuilder {
             }
         }
 
-        // TODO
-        // inputs.sort();
-        // outputs.sort();
+        // Inputs must be lexicographically sorted in their serialised form
```
Comment (thibault-martinez) [inline bee-message/src/payload/transaction/essence/regular.rs]:
```suggestion
        // Inputs must be lexicographically sorted in their serialised form.
```

File: bee-message/src/payload/transaction/essence/regular.rs [hunk 3]
```diff
@@ -234,12 +235,25 @@ impl RegularEssenceBuilder {
...
+            input.pack(&mut packed_input).and(Ok(packed_input))
+        }).collect();
 
-        self.inputs.sort();
-        self.outputs.sort();
+        if !is_sorted(packed_inputs?.iter()) {
+            return Err(Error::TransactionInputsNotSorted);
+        }
+
+        // Outputs must be lexicographically sorted in their serialised form
```
Comment (thibault-martinez) [inline bee-message/src/payload/transaction/essence/regular.rs]:
```suggestion
        // Outputs must be lexicographically sorted in their serialised form.
```

File: bee-message/src/payload/transaction/essence/regular.rs (hunk 1 above)
Comment (Adam-Gleave) [reply bee-message/src/payload/transaction/essence/regular.rs]: Didn't see that one there, makes things much cleaner!

File: bee-message/src/payload/transaction/essence/regular.rs (hunk 2 above)
Comment (Adam-Gleave) [reply bee-message/src/payload/transaction/essence/regular.rs]: Addressed in latest commit

File: bee-message/src/payload/transaction/essence/regular.rs (hunk 3 above)
Comment (Adam-Gleave) [reply bee-message/src/payload/transaction/essence/regular.rs]: Addressed in latest commit

**Review by thibault-martinez — APPROVED** [review]

**Pull request merged**